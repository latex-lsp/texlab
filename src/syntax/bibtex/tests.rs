use insta::assert_debug_snapshot;

use super::{parse, SyntaxNode};

fn setup(input: &str) -> SyntaxNode {
    SyntaxNode::new_root(parse(input))
}

#[test]
fn test_preamble() {
    assert_debug_snapshot!(setup(r#"@preamble{"Hello World!"}"#));
}

#[test]
fn test_comment() {
    assert_debug_snapshot!(setup(
        r#"@comment{aksin,
    author       = {Aks{\i}n, {\"O}zge and T{\"u}rkmen, Hayati and Artok, Levent
                    and {\c{C}}etinkaya, Bekir and Ni, Chaoying and
                    B{\"u}y{\"u}kg{\"u}ng{\"o}r, Orhan and {\"O}zkal, Erhan},
    title        = {Effect of immobilization on catalytic characteristics of
                    saturated {Pd-N}-heterocyclic carbenes in {Mizoroki-Heck}
                    reactions},
    journaltitle = jomch,
    date         = 2006,
    volume       = 691,
    number       = 13,
    pages        = {3027-3036},
    indextitle   = {Effect of immobilization on catalytic characteristics},
}"#
    ));
}

#[test]
fn test_entry_colon() {
    assert_debug_snapshot!(setup(
        r#"@article{foo:2019,
    author = {Foo Bar},
    title = {Baz Qux},
    year = {2019},
}"#
    ));
}

#[test]
fn test_biblatex_examples_001() {
    assert_debug_snapshot!(setup(
        r#"@incollection{westfahl:space,
    author       = {Westfahl, Gary},
    title        = {The True Frontier},
    subtitle     = {Confronting and Avoiding the Realities of Space in {American}
                    Science Fiction Films},
    pages        = {55-65},
    crossref     = {westfahl:frontier},
    langid       = {english},
    langidopts   = {variant=american},
    indextitle   = {True Frontier, The},
    annotation   = {A cross-referenced article from a \texttt{collection}. This is
                    an \texttt{incollection} entry with a \texttt{crossref}
                    field. Note the \texttt{subtitle} and \texttt{indextitle}
                    fields},
}

% booktitle and booksubtitle are only needed for BibTeX's less sophisticated
% inheritance set-up to make sure westfahl:space shows correctly.
% With Biber they are not needed.
@collection{westfahl:frontier,
    editor       = {Westfahl, Gary},
    title        = {Space and Beyond},
    date         = 2000,
    subtitle     = {The Frontier Theme in Science Fiction},
    publisher    = {Greenwood},
    location     = {Westport, Conn. and London},
    langid       = {english},
    langidopts   = {variant=american},
    booktitle    = {Space and Beyond},
    booksubtitle = {The Frontier Theme in Science Fiction},
    annotation   = {This is a \texttt{collection} entry. Note the format of the
                    \texttt{location} field as well as the \texttt{subtitle}
                    field},
}"#
    ));
}

#[test]
fn test_biblatex_examples_002() {
    assert_debug_snapshot!(setup(
        r#"@string{jomch   = {J.~Organomet. Chem.}}

@article{aksin,
    author       = {Aks{\i}n, {\"O}zge and T{\"u}rkmen, Hayati and Artok, Levent
                    and {\c{C}}etinkaya, Bekir and Ni, Chaoying and
                    B{\"u}y{\"u}kg{\"u}ng{\"o}r, Orhan and {\"O}zkal, Erhan},
    title        = {Effect of immobilization on catalytic characteristics of
                    saturated {Pd-N}-heterocyclic carbenes in {Mizoroki-Heck}
                    reactions},
    journaltitle = jomch,
    date         = 2006,
    volume       = 691,
    number       = 13,
    pages        = {3027-3036},
    indextitle   = {Effect of immobilization on catalytic characteristics},
}"#
    ));
}

#[test]
fn test_biblatex_examples_003() {
    assert_debug_snapshot!(setup(
        r#"@article{angenendt,
    author       = {Angenendt, Arnold},
    title        = {In Honore Salvatoris~-- Vom Sinn und Unsinn der
                    Patrozinienkunde},
    journaltitle = {Revue d'Histoire Eccl{\'e}siastique},
    date         = 2002,
    volume       = 97,
    pages        = {431--456, 791--823},
    langid       = {german},
    indextitle   = {In Honore Salvatoris},
    shorttitle   = {In Honore Salvatoris},
    annotation   = {A German article in a French journal. Apart from that, a
                    typical \texttt{article} entry. Note the \texttt{indextitle}
                    field},
}"#
    ));
}
