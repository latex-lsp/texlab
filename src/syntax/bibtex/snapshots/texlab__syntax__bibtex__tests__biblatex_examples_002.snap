---
source: src/syntax/bibtex/tests.rs
assertion_line: 87
expression: "setup(r#\"@string{jomch   = {J.~Organomet. Chem.}}\n\n@article{aksin,\n    author       = {Aks{\\i}n, {\\\"O}zge and T{\\\"u}rkmen, Hayati and Artok, Levent\n                    and {\\c{C}}etinkaya, Bekir and Ni, Chaoying and\n                    B{\\\"u}y{\\\"u}kg{\\\"u}ng{\\\"o}r, Orhan and {\\\"O}zkal, Erhan},\n    title        = {Effect of immobilization on catalytic characteristics of\n                    saturated {Pd-N}-heterocyclic carbenes in {Mizoroki-Heck}\n                    reactions},\n    journaltitle = jomch,\n    date         = 2006,\n    volume       = 691,\n    number       = 13,\n    pages        = {3027-3036},\n    indextitle   = {Effect of immobilization on catalytic characteristics},\n}\"#)"
---
ROOT@0..679
  STRING@0..42
    TYPE@0..7 "@string"
    L_CURLY@7..8 "{"
    KEY@8..13 "jomch"
    WHITESPACE@13..16 "   "
    EQ@16..17 "="
    WHITESPACE@17..18 " "
    CURLY_GROUP@18..39
      L_CURLY@18..19 "{"
      WORD@19..32 "J.~Organomet."
      WHITESPACE@32..33 " "
      WORD@33..38 "Chem."
      R_CURLY@38..39 "}"
    R_CURLY@39..40 "}"
    WHITESPACE@40..42 "\n\n"
  ENTRY@42..679
    TYPE@42..50 "@article"
    L_CURLY@50..51 "{"
    KEY@51..56 "aksin"
    COMMA@56..57 ","
    WHITESPACE@57..62 "\n    "
    FIELD@62..289
      KEY@62..68 "author"
      WHITESPACE@68..75 "       "
      EQ@75..76 "="
      WHITESPACE@76..77 " "
      CURLY_GROUP@77..283
        L_CURLY@77..78 "{"
        WORD@78..81 "Aks"
        CURLY_GROUP@81..85
          L_CURLY@81..82 "{"
          COMMAND_NAME@82..84 "\\i"
          R_CURLY@84..85 "}"
        WORD@85..86 "n"
        COMMA@86..87 ","
        WHITESPACE@87..88 " "
        CURLY_GROUP@88..93
          L_CURLY@88..89 "{"
          COMMAND_NAME@89..91 "\\\""
          WORD@91..92 "O"
          R_CURLY@92..93 "}"
        WORD@93..96 "zge"
        WHITESPACE@96..97 " "
        WORD@97..100 "and"
        WHITESPACE@100..101 " "
        WORD@101..102 "T"
        CURLY_GROUP@102..107
          L_CURLY@102..103 "{"
          COMMAND_NAME@103..105 "\\\""
          WORD@105..106 "u"
          R_CURLY@106..107 "}"
        WORD@107..112 "rkmen"
        COMMA@112..113 ","
        WHITESPACE@113..114 " "
        WORD@114..120 "Hayati"
        WHITESPACE@120..121 " "
        WORD@121..124 "and"
        WHITESPACE@124..125 " "
        WORD@125..130 "Artok"
        COMMA@130..131 ","
        WHITESPACE@131..132 " "
        WORD@132..138 "Levent"
        WHITESPACE@138..159 "\n                    "
        WORD@159..162 "and"
        WHITESPACE@162..163 " "
        CURLY_GROUP@163..170
          L_CURLY@163..164 "{"
          COMMAND_NAME@164..166 "\\c"
          CURLY_GROUP@166..169
            L_CURLY@166..167 "{"
            WORD@167..168 "C"
            R_CURLY@168..169 "}"
          R_CURLY@169..170 "}"
        WORD@170..178 "etinkaya"
        COMMA@178..179 ","
        WHITESPACE@179..180 " "
        WORD@180..185 "Bekir"
        WHITESPACE@185..186 " "
        WORD@186..189 "and"
        WHITESPACE@189..190 " "
        WORD@190..192 "Ni"
        COMMA@192..193 ","
        WHITESPACE@193..194 " "
        WORD@194..202 "Chaoying"
        WHITESPACE@202..203 " "
        WORD@203..206 "and"
        WHITESPACE@206..227 "\n                    "
        WORD@227..228 "B"
        CURLY_GROUP@228..233
          L_CURLY@228..229 "{"
          COMMAND_NAME@229..231 "\\\""
          WORD@231..232 "u"
          R_CURLY@232..233 "}"
        WORD@233..234 "y"
        CURLY_GROUP@234..239
          L_CURLY@234..235 "{"
          COMMAND_NAME@235..237 "\\\""
          WORD@237..238 "u"
          R_CURLY@238..239 "}"
        WORD@239..241 "kg"
        CURLY_GROUP@241..246
          L_CURLY@241..242 "{"
          COMMAND_NAME@242..244 "\\\""
          WORD@244..245 "u"
          R_CURLY@245..246 "}"
        WORD@246..248 "ng"
        CURLY_GROUP@248..253
          L_CURLY@248..249 "{"
          COMMAND_NAME@249..251 "\\\""
          WORD@251..252 "o"
          R_CURLY@252..253 "}"
        WORD@253..254 "r"
        COMMA@254..255 ","
        WHITESPACE@255..256 " "
        WORD@256..261 "Orhan"
        WHITESPACE@261..262 " "
        WORD@262..265 "and"
        WHITESPACE@265..266 " "
        CURLY_GROUP@266..271
          L_CURLY@266..267 "{"
          COMMAND_NAME@267..269 "\\\""
          WORD@269..270 "O"
          R_CURLY@270..271 "}"
        WORD@271..275 "zkal"
        COMMA@275..276 ","
        WHITESPACE@276..277 " "
        WORD@277..282 "Erhan"
        R_CURLY@282..283 "}"
      COMMA@283..284 ","
      WHITESPACE@284..289 "\n    "
    FIELD@289..476
      KEY@289..294 "title"
      WHITESPACE@294..302 "        "
      EQ@302..303 "="
      WHITESPACE@303..304 " "
      CURLY_GROUP@304..470
        L_CURLY@304..305 "{"
        WORD@305..311 "Effect"
        WHITESPACE@311..312 " "
        WORD@312..314 "of"
        WHITESPACE@314..315 " "
        WORD@315..329 "immobilization"
        WHITESPACE@329..330 " "
        WORD@330..332 "on"
        WHITESPACE@332..333 " "
        WORD@333..342 "catalytic"
        WHITESPACE@342..343 " "
        WORD@343..358 "characteristics"
        WHITESPACE@358..359 " "
        WORD@359..361 "of"
        WHITESPACE@361..382 "\n                    "
        WORD@382..391 "saturated"
        WHITESPACE@391..392 " "
        CURLY_GROUP@392..398
          L_CURLY@392..393 "{"
          WORD@393..397 "Pd-N"
          R_CURLY@397..398 "}"
        WORD@398..411 "-heterocyclic"
        WHITESPACE@411..412 " "
        WORD@412..420 "carbenes"
        WHITESPACE@420..421 " "
        WORD@421..423 "in"
        WHITESPACE@423..424 " "
        CURLY_GROUP@424..460
          L_CURLY@424..425 "{"
          WORD@425..438 "Mizoroki-Heck"
          R_CURLY@438..439 "}"
          WHITESPACE@439..460 "\n                    "
        WORD@460..469 "reactions"
        R_CURLY@469..470 "}"
      COMMA@470..471 ","
      WHITESPACE@471..476 "\n    "
    FIELD@476..502
      KEY@476..488 "journaltitle"
      WHITESPACE@488..489 " "
      EQ@489..490 "="
      WHITESPACE@490..491 " "
      LITERAL@491..496
        WORD@491..496 "jomch"
      COMMA@496..497 ","
      WHITESPACE@497..502 "\n    "
    FIELD@502..527
      KEY@502..506 "date"
      WHITESPACE@506..515 "         "
      EQ@515..516 "="
      WHITESPACE@516..517 " "
      LITERAL@517..521
        INTEGER@517..521 "2006"
      COMMA@521..522 ","
      WHITESPACE@522..527 "\n    "
    FIELD@527..551
      KEY@527..533 "volume"
      WHITESPACE@533..540 "       "
      EQ@540..541 "="
      WHITESPACE@541..542 " "
      LITERAL@542..545
        INTEGER@542..545 "691"
      COMMA@545..546 ","
      WHITESPACE@546..551 "\n    "
    FIELD@551..574
      KEY@551..557 "number"
      WHITESPACE@557..564 "       "
      EQ@564..565 "="
      WHITESPACE@565..566 " "
      LITERAL@566..568
        INTEGER@566..568 "13"
      COMMA@568..569 ","
      WHITESPACE@569..574 "\n    "
    FIELD@574..606
      KEY@574..579 "pages"
      WHITESPACE@579..587 "        "
      EQ@587..588 "="
      WHITESPACE@588..589 " "
      CURLY_GROUP@589..600
        L_CURLY@589..590 "{"
        WORD@590..599 "3027-3036"
        R_CURLY@599..600 "}"
      COMMA@600..601 ","
      WHITESPACE@601..606 "\n    "
    FIELD@606..678
      KEY@606..616 "indextitle"
      WHITESPACE@616..619 "   "
      EQ@619..620 "="
      WHITESPACE@620..621 " "
      CURLY_GROUP@621..676
        L_CURLY@621..622 "{"
        WORD@622..628 "Effect"
        WHITESPACE@628..629 " "
        WORD@629..631 "of"
        WHITESPACE@631..632 " "
        WORD@632..646 "immobilization"
        WHITESPACE@646..647 " "
        WORD@647..649 "on"
        WHITESPACE@649..650 " "
        WORD@650..659 "catalytic"
        WHITESPACE@659..660 " "
        WORD@660..675 "characteristics"
        R_CURLY@675..676 "}"
      COMMA@676..677 ","
      WHITESPACE@677..678 "\n"
    R_CURLY@678..679 "}"

